crates/sui-indexer-alt-framework/src/ingestion/streaming_service.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use async_trait::async_trait;
+use futures::{stream::Peekable, StreamExt};
+use sui_rpc::{
+    field::{FieldMask, FieldMaskUtil},
+    proto::sui::rpc::v2::{
+        subscription_service_client::SubscriptionServiceClient, SubscribeCheckpointsRequest,
+        SubscribeCheckpointsResponse,
+    },
+};
+use tonic::{Status, Streaming};
+
+use crate::ingestion::error::{Error, Result};
+use crate::types::full_checkpoint_content::CheckpointData;
+
+#[async_trait]
+pub trait StreamingService {
+    async fn start_streaming(&mut self) -> Result<()>;
+    async fn next_checkpoint(&mut self) -> Result<CheckpointData>;
+    async fn peek_next_checkpoint(&mut self) -> Result<u64>;
+}
```

Comment:
Something like this:

```suggestion
#[async_trait]
pub trait StreamingService {
    type Stream: Stream;
    async fn connect(&mut self) -> Result<Stream>;
}

#[async_trait]
trait PeekableExt<T>: Stream<Item = T> {
    async fn peek(&self) -> Option<&T>;
}
```